File: tests/cuda/lua/test_signal_radar.lua
-- Signal radar CUDA binding tests
-- Run with:
--   LUA_PATH="bindings/lua/?/init.lua;;" LUA_CPATH="build/bindings/lua/?.so;;" \
--   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda \
--   ~/.luarocks/bin/busted tests/cuda/lua/test_signal_radar.lua

local ok_ins, ins = pcall(require, "_insight")
if not ok_ins then
  print("SKIP: _insight module not available")
  os.exit(0)
end

local ok_gpu = pcall(function()
  ins.init()
end)
if not ok_gpu then
  print("SKIP: GPU backend not available")
  os.exit(0)
end

ins.set_device(ins.GPUPlace(0))

describe("Signal Radar CUDA Tests", function()
  it("pulse_compression", function()
    local n = 64
    local tpl_t = {}
    local sig_t = {}
    for i = 1, n do
      tpl_t[i] = math.sin(2 * math.pi * i / n)
      sig_t[i] = tpl_t[i]
    end
    for i = n + 1, 2 * n do
      sig_t[i] = 0.0
    end
    local tpl = ins.from_table(tpl_t):to(ins.GPUPlace(0))
    local sig = ins.from_table(sig_t):to(ins.GPUPlace(0))
    local result = ins.signal.pulse_compression(sig, tpl)
    assert.is_not_nil(result)
    assert.is_true(result.numel > 0)
  end)

  it("pulse_doppler", function()
    local rows = {}
    for r = 1, 16 do
      local row = {}
      for c = 1, 32 do
        row[c] = math.sin(2 * math.pi * r * c / 512)
      end
      rows[r] = row
    end
    local data = ins.from_table(rows):to(ins.GPUPlace(0))
    local result = ins.signal.pulse_doppler(data, 16)
    assert.is_not_nil(result)
    assert.is_true(result.numel > 0)
  end)

  it("cfar_alpha", function()
    local alpha = ins.signal.cfar_alpha(20, 1e-3)
    assert.is_not_nil(alpha)
    assert.is_true(alpha > 0)
  end)

  it("cfar_alpha_different_pfa", function()
    local alpha1 = ins.signal.cfar_alpha(20, 1e-2)
    local alpha2 = ins.signal.cfar_alpha(20, 1e-6)
    assert.is_true(alpha1 < alpha2)
  end)

  it("ca_cfar", function()
    local t = {}
    for i = 1, 100 do
      t[i] = 0.0
    end
    t[30] = 10.0
    t[70] = 10.0
    local data = ins.from_table(t):to(ins.GPUPlace(0))
    local result = ins.signal.ca_cfar(data, 10, 2, 1e-2)
    assert.is_not_nil(result)
    assert.are.equal(100, result.numel)
  end)

  it("mvdr", function()
    local rows = {}
    for r = 1, 4 do
      local row = {}
      for c = 1, 32 do
        row[c] = math.sin(2 * math.pi * r * c / 128)
      end
      rows[r] = row
    end
    local x = ins.from_table(rows):to(ins.GPUPlace(0))
    local sv = ins.from_table({ 1.0, 1.0, 1.0, 1.0 }):to(ins.GPUPlace(0))
    local result = ins.signal.mvdr(x, sv)
    assert.is_not_nil(result)
    assert.is_true(result.numel > 0)
  end)

  it("ambgfun", function()
    local t = {}
    for i = 1, 64 do
      t[i] = math.sin(2 * math.pi * i / 64)
    end
    local x = ins.from_table(t):to(ins.GPUPlace(0))
    local result = ins.signal.ambgfun(x, 32, 32)
    assert.is_not_nil(result)
    assert.is_true(result.numel > 0)
  end)
end)

File: tests/cuda/julia/test_signal_radar.jl
# Signal radar CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_signal_radar.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

# ============================================================================
# Radar Signal Processing
# ============================================================================
println("=== Signal Radar ===")

# pulse_compression
n = 64
tpl_data = [sin(2 * pi * i / n) for i in 1:n]
sig_data = vcat(tpl_data, zeros(n))
tpl = Insight.from_data(tpl_data)
tpl_gpu = Insight.to_array(tpl, Insight.GPUPlace(0))
sig = Insight.from_data(sig_data)
sig_gpu = Insight.to_array(sig, Insight.GPUPlace(0))
result = Insight.pulse_compression(sig_gpu, tpl_gpu)
check("pulse_compression", Insight.numel(result) > 0)

# pulse_compression normalized
result = Insight.pulse_compression(sig_gpu, tpl_gpu, normalize=true)
check("pulse_compression_norm", Insight.numel(result) > 0)

# pulse_doppler
data_rows = [sin(2 * pi * r * c / 512) for r in 1:16, c in 1:32]
data_flat = vec(data_rows')
data_arr = Insight.from_data(collect(data_flat))
data_2d = Insight.reshape(data_arr, [32, 16])
data_gpu = Insight.to_array(data_2d, Insight.GPUPlace(0))
result = Insight.pulse_doppler(data_gpu)
check("pulse_doppler", Insight.numel(result) > 0)

# cfar_alpha
alpha = Insight.cfar_alpha(1e-3, 20)
check("cfar_alpha", alpha > 0.0)

# cfar_alpha different PFA
alpha1 = Insight.cfar_alpha(1e-2, 20)
alpha2 = Insight.cfar_alpha(1e-6, 20)
check("cfar_alpha_pfa", alpha1 < alpha2)

# mvdr
x_data = [sin(2 * pi * r * c / 128) for r in 1:4, c in 1:32]
x_flat = vec(x_data')
x_arr = Insight.from_data(collect(x_flat))
x_2d = Insight.reshape(x_arr, [32, 4])
x_gpu = Insight.to_array(x_2d, Insight.GPUPlace(0))
sv = Insight.from_data([1.0, 1.0, 1.0, 1.0])
sv_gpu = Insight.to_array(sv, Insight.GPUPlace(0))
result = Insight.mvdr(x_gpu, sv_gpu)
check("mvdr", Insight.numel(result) > 0)

# mvdr with different steering
sv2 = Insight.from_data([1.0, 0.5, 0.5, 1.0])
sv2_gpu = Insight.to_array(sv2, Insight.GPUPlace(0))
result2 = Insight.mvdr(x_gpu, sv2_gpu)
check("mvdr_steering", Insight.numel(result2) > 0)

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
